jlm/llvm/opt/alias-analyses/RegionAwareModRefSummarizer.hpp

```diff
@@ -13,30 +14,40 @@ namespace jlm::llvm::aa
 {
 
 class RegionAwareModRefSummary;
-class RegionSummary;
+using ModRefSetIndex = uint32_t;
 
 /** \brief Region-aware mod/ref summarizer
  *
- * The key idea of the region-aware memory mod/ref summarizer is to only provide memory locations
- * for a structural node that are actually utilized within its regions. This ensures that no
+ * The key idea of the region-aware memory mod/ref summarizer is to only route memory locations
+ * into structural nodes if the memory is actually used within its regions. This ensures that no
  * superfluous states will be routed through structural nodes and renders them independent if they
  * do not reference the same memory location. The region-aware analysis proceeds as follows:
  *
  * 1. Call Graph Creation: creates a call graph by looking at all call operations.
  * This graph includes calls to external functions, and calls from external functions.
  * Each function is assigned to a strongly connected component.
  *
- * 2. These strongly connected components are then visited in reverse topological order.
- * This process is called Annotation, and creates summaries for all regions and calls.
- * These summaries contain the sets of memory locations that may be read from or written to
- * inside the region or by the target of the call.
+ * 2. Find allocas that are dead in each SCC:
+ * For each SCC in the call graph, only allocas defined within the SCC,
+ * or within one of its predecessors, can be live.
+ * All other allocas are placed in the DeadAllocasInScc lists.
  *
- * Since a call may target another function in the same SCC that has not been annotated yet,
- * a flag is set on all RegionSummary and CallSummary instances that may contain recursion.
+ * 3. Simple Alloca Set Creation: An alloca is "simple" if its address is never stored to
+ * any memory location, except for other simple allocas.
+ * The PointsToGraph is used to determine which allocas are simple.
  *
- * 3. Propagation: Once all functions have been annotated, we go back to all summaries that
- * were flagged as possibly containing recursion. Their sets of memory locations are expanded to
- * include all memory locations that may be affected by any function inside the SCC.
+ * 4. Create sets of non-reentrant allocas for each region.
+ * The requirements are:
+ *  - the alloca must be simple
+ *  - the alloca must not be reachable from any of the region's arguments,
```

Comment:
Should this not be: "reachable from any of the function's arguments"?

Reply:
No, allocas can also (rarely) be defined inside subregions, in which case we use that region to detect it